enhancements/network/ovs-dynamic-cpu-pinning.md

```diff
@@ -0,0 +1,205 @@
+```
+
+### API Extensions
+
+N/A
+
+### Implementation Details/Notes/Constraints
+
+#### CPU aligning
+
+CPU alignment is implemented using system calls SYS_SCHED_SETAFFINITY and SYS_SCHED_GETAFFINITY.
+Pods from `ovnkube-node` daemonset run in the Burstable Quality of Service (QoS) class, hence their CPU 
+affinities are managed by the [kubelet cpu manager](https://kubernetes.io/docs/tasks/administer-cluster/cpu-management-policies/).
+
+#### `/etc/openvswitchd/enable_cpu_pinning` file
+
+The file is used to activate the feature on ovnkube-node according to the following logic:
+a. If the file is missing OR the file file is empty, then the feature is turned `OFF`.
+b. If the file exists and its content is not an empty string, then the feature is turned `ON`.
+c. `ovnkube-node` responds to the file modifications (deletion or content update)
+  at runtime, without the need for a restart.
+
+#### Must-Gather enhancement
+
+As the feature is enabled via a fail on the host filesystem, it's useful to
+add `/etc/openvswitchd/enable_cpu_pinning` file to the [gather_network_logs](https://github.com/openshift/must-gather/blob/master/collection-scripts/gather_network_logs) script.
+
+### Risks and Mitigations
+
+|Risk|Mitigation|
+|---|---|
+|After NTO enabled the feature, a network performance decrease occurred in some cluster nodes.|The cluster administrator can turn off the feature on specific nodes by connecting through SSH and deleting the activation file. An overriding MachineConfig can be created to achieve the same goal.|
+
+
+### Drawbacks
+
+This enhancement breaks the hard division between `kube-reserved` and `system-reserved` CPU pools in a cluster with `cpuManagerPolicy = static`.
+
+
+## Design Details
+
+### Test Plan
+
+Following scenario should be covered by automated tests:
+
+_Given_ an OpenShift cluster with a node N with 10 CPUs (0,1,...,9), configured with 
+- `cpuManagerPolicy = static`
+- `system-reserved = 0,1,2,3`
+
+_And_ a running pod with `Guaranteed` QoS running on cpu 4,5
+
+_When_ the administrator creates a the file `/etc/openvswitch/enable_cpu_pinning` on N's host filesystem
+
+_Then_ `ovs-switchd` and `ovsdb-server` processes on N has CPU affinity equal to `6,7,8,9`.
+
+
+### Graduation Criteria
+#### Dev Preview -> Tech Preview
+#### Tech Preview -> GA
+#### Removing a deprecated feature
+### Upgrade / Downgrade Strategy
+### Version Skew Strategy
+
+### Operational Aspects of API Extensions
+
+No API changes.
+
+#### Failure Modes
+
+No API changes.
+
+#### Support Procedures
+
+The following MachineConfig resource disable the feature
+for every node with `role = worker`:
+
+```yaml
+apiVersion: machineconfiguration.openshift.io/v1
+kind: MachineConfig
+metadata:
+  labels:
+    machineconfiguration.openshift.io/role: worker
+  name: 99-disable-ovs-cpu-pinning
+spec:
+  config:
+    ignition:
+      version: 3.2.0
+    storage:
+      files:
+      - contents:
+          # Empty file disable the feature
```

Comment:
earlier in the enhancement you say the contents are irrelevant and only the presence of the file determines enablement.

With this ignition is it making sure the file is removed or creating it with empty contents?

Reply:
I corrected the document: if the file is empty, the feature is turned off. AFAIK, a MachineConfig can't delete a file created by another MC.

Reply:
Indeed, deletion is hard and so the override can replace the file with an empty content to disable the feature. The typical use case should put an explanation text into the file to make sure administrators understand the purpose when they look at it (and that satisfies the non-empty for enablement condition).